lib/charms/loki_k8s/v1/loki_push_api.py

```diff
@@ -2313,6 +2395,155 @@ def _containers(self) -> Dict[str, Container]:
+class LogForwarder(Object):
+    """Forward the standard outputs of all workloads operated by a charm to one or multiple Loki endpoints."""
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        *,
+        relation_name: str = DEFAULT_RELATION_NAME,
+        loki_endpoints_getter: Optional[Callable[[Relation], List[str]]] = None,
+    ):
+        super().__init__(charm, relation_name)
+        self._charm = charm
+        self._relation_name = relation_name
+        self._topology = JujuTopology.from_charm(charm)
+        self._endpoints_getter: Callable = (
+            self._extract_urls if loki_endpoints_getter is None else loki_endpoints_getter
+        )
+
+        on_relation = self._charm.on[self._relation_name]
+        self.framework.observe(on_relation.relation_joined, self._on_logging_relation_joined)
+        self.framework.observe(on_relation.relation_changed, self._on_logging_relation_changed)
+        self.framework.observe(on_relation.relation_departed, self._on_logging_relation_departed)
+        self.framework.observe(on_relation.relation_broken, self._on_logging_relation_broken)
+
+    def _on_logging_relation_joined(self, _):
+        self._enable_logging()
+
+    def _on_logging_relation_changed(self, _):
+        self._enable_logging()
+
+    def _on_logging_relation_departed(self, event: RelationDepartedEvent):
+        self._disable_logging(unit=event.unit)
+
+    def _on_logging_relation_broken(self, event: RelationBrokenEvent):
+        for unit in event.relation.units:
+            if unit.app == self._charm.app:
+                continue
+            self._disable_logging(unit=unit)
+
+    def _build_log_target(self, unit_name, endpoint, enabled=False):
+        """Build a log target for the log forwarding Pebble layer."""
+        services_value = ["all"] if enabled else ["-all"]
+
+        log_target = {
+            "override": "merge",
+            "services": services_value,
+        }
+        if enabled:
+            log_target.update(
+                {
+                    "type": "loki",
+                    "location": endpoint,
+                    "labels": {
+                        "product": "Juju",
+                        "charm": self._topology._charm_name,
+                        "juju_model": self._topology._model,
+                        "juju_model_uuid": self._topology._model_uuid,
+                        "juju_application": self._topology._application,
+                        "juju_unit": self._topology._unit,
+                    },
+                }
+            )
+
+        return {unit_name: log_target}
+
+    def _build_log_targets(self, endpoints: Optional[Dict[str, str]], enable=False):
+        """Build the targets for the log forwarding Pebble layer."""
+        targets = {}
+        if endpoints:
+            for unit_name, endpoint in endpoints.items():
+                targets.update(self._build_log_target(unit_name, endpoint, enable))
+        return targets
```

Comment:
I would remove one indentation level:

```suggestion
        targets = {}
        if not endpoints:
            return targets

        for unit_name, endpoint in endpoints.items():
            targets.update(self._build_log_target(unit_name, endpoint, enable))
        return targets
```